Bluetooth: controller: Extended Scanning with duration and period

dagbja opened the pull request:
Start/stop scan according to period and duration params

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

dagbja committed:
Bluetooth: controller: Add periodic scanning

Start/stop scan according to period and duration
parameters.

Signed-off-by: Dag Bjarvin <Dag.Bjarvin@nordicsemi.no>

dagbja commented:
hci.c, line 1916,1917 is temporary code to be able to test with shell. Will be removed later.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -131,6 +131,16 @@ uint8_t ll_scan_enable(uint8_t enable)
+	scan->lll.period = (uint64_t)period * (1280000U / 625U);
+	scan->lll.duration = (uint64_t)duration * (10000U / 625U);
+
+	if (scan->lll.duration && scan->lll.period
```
This validation are only required for controller-only build, use conditional compile using CONFIG_BT_CTLR_PARAM_CHECK

cvinayak replied:
Please valid the parameters, before use. Move validation before line 134.

dagbja replied:
Fixed

dagbja replied:
Just remembered why it was after. We need to scale period and duration to comare them. Moved the check back after the scaling.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -482,6 +492,19 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
 	struct ll_scan_set *scan = param;
 	uint32_t ret;
 	uint8_t ref;
+	struct lll_scan *lll;
```
This is incorrect, you have to design a "periodic scanning" for a "duration". You are shunting the "scan interval" after "duration" in your implementation here.

dagbja replied:
Yes, I'm shunting it until start of the next period, so it will scan for "duration", restarting every "period".

cvinayak replied:
By shunting, other BLE states/roles are blocked from using the radio air time as scanner is reserving the window every interval, beats the purpose of period and duration feature.

dagbja replied:
Modified code to avoid this

cvinayak commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.h:
```diff
@@ -38,6 +38,9 @@ struct lll_scan {
 
 	uint16_t interval;
 	uint32_t ticks_window;
+	uint32_t period;
```
move fields to ll_scan_set, as lll_scan member are the ones used/modified by LLL.

dagbja replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -480,9 +492,51 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
+			LL_ASSERT((ret == TICKER_STATUS_SUCCESS) ||
+				  (ret == TICKER_STATUS_BUSY));
+		} else {
+			lll->dur_counter += lll->interval;
```
Use countdown, compared to addition operation.

cvinayak replied:
Also, need to take `lazy` value into account.

dagbja replied:
Fixed

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -131,6 +131,18 @@ uint8_t ll_scan_enable(uint8_t enable)
 	}
 
+	scan->lll.period = (uint64_t)period * (1280000U / 625U);
+	scan->lll.duration = (uint64_t)duration * (10000U / 625U);
```
Store in `interval` units and use countdown

dagbja replied:
Fixed

cvinayak committed:
Bluetooth: controller: Cond. compile related to extended scanning

Refactoring of conditional compilation related to extended
scanning with duration and period.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix scanning flags reset

Fix scanning state and is_adv_ind flags to be common to both
isr_done and isr_window code path.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Move scanning channel set implementation

Move the implementation to set scanning channel to
isr_window function where the next window is started,
instead of being at the abort of the previous continuous
window.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Fix scan done generation

Fix scan done to be generated when the window is closed
both under abort and graceful window close.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Revert ticker status check

Revert the added ticker status busy check.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
Bluetooth: controller: Extended Scan with duration and period

Added implementation for Extended Scanning with duration
and period parameters.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
tests: Bluetooth: bsim: Extended scanning with duration

Added test to verify Extended Scanning with duration
parameter.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
@Thalley This PR illustrates changes spanning across controller's HCI->ULL->LLL and back to HCI generation of event. Reviewing, debugging its controller path, and testing the implementation here will give understanding of one of the design pattern followed. Most new work would be repetition of these patterns.

@wopu-ot and @kruithofa the PR is a simple implementation of duration and scan period, in units of scan interval, hence the scan period will drift, hope this is acceptable as a initial support? accuracy can be improved by maintaining a `remainder` to nudge the scan period, this should be done in a subsequent PR.

Also, enabling an already enabled scanning needs to update/reset the filter, duration and scan period; I will send follow up commits.

Please do review as appropriate.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -164,68 +164,70 @@ enum node_rx_type {
 /* Following proprietary defines must be at end of enum range */
 #if defined(CONFIG_BT_CTLR_USER_EXT)
-	NODE_RX_TYPE_USER_START = 0x18,
+	NODE_RX_TYPE_USER_START = 0x1a,
```
This value is incremented by 2 but only one type was inserted above. But if I understand correctly, this change actually fixes a lurking issue.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_adv.c:
```diff
@@ -1511,13 +1511,14 @@ void ull_adv_done(struct node_rx_event_done *done)
+	LL_ASSERT(handle < BT_CTLR_ADV_SET);
+
 	rx_hdr->type = NODE_RX_TYPE_EXT_ADV_TERMINATE;
+	rx_hdr->handle = handle;
 	rx_hdr->rx_ftr.param_adv_term.conn_handle = 0xffff;
 	rx_hdr->rx_ftr.param_adv_term.num_events = adv->event_counter;
 
-	handle = ull_adv_handle_get(adv);
-	LL_ASSERT(handle < BT_CTLR_ADV_SET);
-
```
These changes seem to be unrelated to scanning. I would prefer to move them to a separate PR (which then probably will be trivial to review and get merged).

wopu-ot commented on tests/bluetooth/bsim_bt/bsim_test_advx/src/main.c:
```diff
@@ -921,6 +929,39 @@ static void test_scanx_main(void)
+	printk("success.\n");
+
+	printk("Waiting...");
+	while (!is_scan_timeout) {
```
I think it would be better to wait for the set duration and then check the value of is_scan_timeout to avoid hanging if the termination fails.

cvinayak replied:
ok.. I will use k_sleep for 1010 ms and then check if `is_scan_timeout` has been set.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
+			/* The alloc here used for ext scan termination event */
+			link_scan_term = ll_rx_link_alloc();
+			if (!link_scan_term) {
+				/* TODO: figure out right return value */
+				return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
```
I would argue that `BT_HCI_ERR_MEM_CAPACITY_EXCEEDED` sounds like a proper return value, and the TODO (both this and the one below) should be removed.

cvinayak replied:
What about insufficient resources?

Thalley replied:
That would also be a good alternative. However, it seems that `BT_HCI_ERR_MEM_CAPACITY_EXCEEDED` is typically returned when `ll_rx_xxx_alloc()` fails. See also `ull_adv.c:740`.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_slave.c:
```diff
@@ -230,8 +230,8 @@ void ull_slave_setup(memq_link_t *link, struct node_rx_hdr *rx,
 
-		rx->handle = ull_adv_handle_get(adv);
 		rx->type = NODE_RX_TYPE_EXT_ADV_TERMINATE;
+		rx->handle = ull_adv_handle_get(adv);
```
Seems like an unnecessary change.

cvinayak replied:
This is make it consistent, to set the handle after the type.

cvinayak committed:
Bluetooth: controller: Fix the ll_scan_enable signature

Fix the ll_scan_enable interface signature, place duration
before period parameter to reflect the order as in the HCI
LE Set Extended Scan Enable command.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
+
+			scan->duration_lazy = 0U;
+
+			/* The alloc here used for ext scan termination event */
```
The pool is owned by the thread context, and is to be allocated/released in thread context, hence the reservation.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
+
+			scan->duration_lazy = ((uint32_t)period * 1280000U /
+					       625U) / scan->lll.interval;
+			scan->duration_lazy -= lll->duration_reload;
```
duration_lazy is directly supplied to ticker_update to stop scan windows from end of duration until scan period ends.

wopu-ot replied:
I would find it more intuitive to store the period and subtract the duration when updating the ticker, but if the other reviewers are okay with how it is done now, I will not object.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
+				return BT_HCI_ERR_INVALID_PARAM;
+			}
+
+			scan->duration_lazy = ((uint32_t)period * 1280000U /
```
Please suggest a define name.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
+					scan->lll.interval;
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period << 7))) {
```
Suggestion to code change welcome.

wopu-ot replied:
I suggest
`#define EXT_SCAN_PERIOD_UNIT_US 1280000U`
`#define EXT_SCAN_DURATION_UNIT_US 10000U`
...
`... duration >= (uint32_t)period * (EXT_SCAN_PERIOD_UNIT_US / EXT_SCAN_DURATION_UNIT_US) ...`
I am not sure where the best place for the defines would be, but I believe you will find the right place for them.

cvinayak commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -151,9 +161,56 @@ uint8_t ll_scan_enable(uint8_t enable)
+#if defined(CONFIG_BT_CTLR_ADV_EXT)
+	lll = &scan->lll;
+	if (duration) {
+		lll->duration_reload = ((uint32_t)duration * 10000U / 625U) /
```
Will be handle as a whole in https://github.com/zephyrproject-rtos/zephyr/issues/23314

cvinayak committed:
Bluetooth: controller: Use defines in place of magic numbers

Use defines for Extended Scan duration, period and scan
interval units.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak commented:
During testing on nrf52840 target that supported Coded PHY, I discovered issues with terminating scan with two scanning instances. Also, now support for updating scan duration and period of an active scanning instance is added and tested. I will clean up commits and push new commits soon.

Below is the screenshot testing scanning instances:
![image](https://user-images.githubusercontent.com/6350656/96208183-3b8ed380-0f6d-11eb-8c0e-2077bad3d9b7.png)

cvinayak committed:
Bluetooth: controller: Extended Scanning Coded PHY duration and period

Added implementation to support simultaneous 1M and Coded
PHY Extended Scanning duration and period. Also, added
implementation to support update to duration and period of
an active scanning instance.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -164,16 +164,22 @@ uint8_t ll_scan_enable(uint8_t enable)
-		lll->duration_reload = ((uint32_t)duration * 10000U / 625U) /
+		lll->duration_reload = ((uint32_t)duration *
+					EXT_SCAN_DURATION_UNIT_US /
+					SCAN_INTERVAL_UNIT_US) /
```
I wonder if it would be more readable to have a macro for converting this value, so it'd be `lll->duration_reload = CONVERT(duration, interval)` (with a much better name than `CONVERT` of course. Similar macro can be added for the duration calculation below.

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -460,6 +480,88 @@ uint8_t ull_scan_disable(uint8_t handle, struct ll_scan_set *scan)
+{
+	struct ll_scan_set *scan;
+
+	scan = ull_scan_set_get(handle);
+	scan->is_enabled = 0U;
```
Any risk that `ull_scan_set_get` returns NULL here?

Thalley replied:
Seems like there are plenty of other places where there is a NULL check following this get.

cvinayak replied:
I have only added the check in one place, other places the caller uses a constant valid handle value or `SCAN_HANDLE_1M` or `SCAN_HANDLE_PHY_CODED`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -622,41 +728,116 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder,
+			      uint16_t period, struct ll_scan_set **scan,
+			      struct node_rx_pdu **node_rx_scan_term)
+{
+	*scan = ull_scan_set_get(handle);
+	*node_rx_scan_term = (void *)(*scan)->node_rx_scan_term;
```
Missing NULL check for `*scan`?

cvinayak replied:
The caller uses a constant valid handle value of `SCAN_HANDLE_1M` or `SCAN_HANDLE_PHY_CODED`

Thalley commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -622,41 +728,116 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder,
-		return;
+	lll = &scan->lll;
+	if (duration) {
+		lll->duration_reload = ((uint32_t)duration *
+					EXT_SCAN_DURATION_UNIT_US /
+					SCAN_INTERVAL_UNIT_US) /
+					scan->lll.interval;
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period *
+					  EXT_SCAN_PERIOD_UNIT_US/
+					  EXT_SCAN_DURATION_UNIT_US))) {
+				return BT_HCI_ERR_INVALID_PARAM;
+			}
+
+			scan->duration_lazy = ((uint32_t)period *
+					       EXT_SCAN_PERIOD_UNIT_US /
+					       SCAN_INTERVAL_UNIT_US) /
+					      scan->lll.interval;
```
See my other comment regarding conversion macros.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -175,7 +232,32 @@ uint8_t ll_scan_enable(uint8_t enable)
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) &&
```
This is inside an #if defined(CONFIG_BT_CTLR_ADV_EXT) block, no need to check for IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT).

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -129,28 +149,65 @@ uint8_t ll_scan_enable(uint8_t enable)
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT) &&
```
This is inside an #if defined(CONFIG_BT_CTLR_ADV_EXT) block, no need to check for IS_ENABLED(CONFIG_BT_CTLR_ADV_EXT).

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -510,6 +727,170 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
+{
+	*scan = ull_scan_set_get(handle);
+	*node_rx_scan_term = (void *)(*scan)->node_rx_scan_term;
+	return duration && period && (*scan)->lll.duration_reload &&
+	       (*scan)->duration_lazy;
```
I did not find in the spec that old or new duration parameters need to be non-zero for a scan update. But maybe I am overlooking something?

cvinayak replied:
BT Spec v5.2, Vol.4, Part E, Section 7.8.65

> When the Duration and Period parameters are non-zero, the Controller shall scan for the duration of the Duration parameter within a scan period specified by the Period parameter.

and 

> If the HCI_LE_Set_Extended_Scan_Enable command with Enable set to 0x01 is sent while scanning is already enabled, the timers used for duration and period are reset to the new parameter values and a new scan period is started.

For the update to be allowed both duration and period need to be non-zero. As Period parameter specifies 'scan period'.

wopu-ot replied:
Do I understand correctly that you interpret this such that the "new scan period" mentioned in the spec implies that there must be an "old scan period"? The test specification does not seem to test this behavior, btw.

cvinayak replied:
> you interpret this such that the "new scan period" mentioned in the spec implies that there must be an "old scan period"

Yes.

> The test specification does not seem to test this behavior, btw.

I did not get, there is a test that wants to switch from a scanning with non-zero Duration and zero Period to both non-zero and vice versa? Could you point me to the test cases that I can refer to, to make changes accordingly.

wopu-ot replied:
Looks like I did not express myself clearly... I could not find any test case that uses a non-zero period, and only one that uses a non-zero duration. There does not seem to be any test case that updates these parameters. We therefore cannot use the test spec to infer the intended behavior. I do not find the wording in the BT Spec crystal clear, but I think your interpretation is valid.

cvinayak replied:
If the implementation needs to allow updates to scanning irrespective of old duration and period, its is not difficult to fix it based on a requirement/request from users. Additonal implementation is required to release buffers allocated for scan terminate event, if period is updated from zero to non-zero value.

wopu-ot commented on subsys/bluetooth/controller/ll_sw/ull_scan.c:
```diff
@@ -510,6 +727,170 @@ static void ticker_cb(uint32_t ticks_at_expire, uint32_t remainder, uint16_t laz
+		if (period) {
+			if (IS_ENABLED(CONFIG_BT_CTLR_PARAM_CHECK) &&
+			    (duration >= ((uint32_t)period *
+					  EXT_SCAN_PERIOD_UNIT_US/
```
nitpick: space before /

cvinayak committed:
Bluetooth: controller: Use conversion macros for duration and period

Use conversion macros to convert Extended Scan duration and
period to radio event counts.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on subsys/bluetooth/controller/ll_sw/nordic/lll/lll_scan.c:
```diff
@@ -562,9 +572,17 @@ static void isr_window(void *param)
 	isr_common_done(param);
 
+	lll = param;
+	if (++lll->chan == 3U) {
+		lll->chan = 0U;
+	}
```
`3U` seems like a magic number in this context. Hard to tell what's going on here. Consider adding a comment about why comparing to 3, and why it's `++lll->chan` rather than `lll->chan`.

Thalley reviewed (changes requested):
Small request for added readability, otherwise LGTM.

trond-snekvik commented:
This broke the build for non-AE apps: https://buildkite.com/zephyr/zephyr/builds/11604
